File: backend/app/core/rich_text.py
"""Sanitizing the announcement rich text field.

The frontend's composer runs a Tiptap editor through DOMPurify with this
same allow-list (`frontend/src/lib/rich-text.ts`) before it ever reaches
this API. This module exists for the request that skips the composer
entirely — a hand-built call against `/api/v1/announcements` — so the tags
an admin could actually produce with the editor's toolbar are the only
ones that ever land in the database, regardless of caller.
"""
from __future__ import annotations

import re

import bleach

# Kept in lockstep with `ALLOWED_TAGS`/`ALLOWED_ATTR` in
# `frontend/src/lib/rich-text.ts` — the editor's toolbar produces exactly
# these tags and nothing else.
ALLOWED_TAGS = [
    "p",
    "br",
    "strong",
    "em",
    "u",
    "s",
    "h2",
    "h3",
    "ul",
    "ol",
    "li",
    "blockquote",
    "a",
]
ALLOWED_ATTRIBUTES = {"a": ["href", "target", "rel"]}

_TAG_RE = re.compile(r"<[^>]*>")
_WHITESPACE_RE = re.compile(r"\s+")


def sanitize_announcement_html(html: str) -> str:
    """Strips anything outside the editor's own allow-list."""
    return bleach.clean(html, tags=ALLOWED_TAGS, attributes=ALLOWED_ATTRIBUTES, strip=True).strip()


def strip_html_to_text(html: str) -> str:
    """Plain text, for length checks the same way the frontend applies them."""
    return _WHITESPACE_RE.sub(" ", _TAG_RE.sub(" ", html)).strip()

File: frontend/src/lib/rich-text.ts
import DOMPurify from "isomorphic-dompurify";

/**
 * Tags the announcement rich text editor's toolbar can actually produce.
 * Kept as a closed allow-list — rather than trusting whatever Tiptap
 * happens to emit — so a request built by hand (bypassing the editor
 * entirely) can't smuggle in anything beyond what an admin using the UI
 * could have written.
 */
const ALLOWED_TAGS = [
  "p",
  "br",
  "strong",
  "em",
  "u",
  "s",
  "h2",
  "h3",
  "ul",
  "ol",
  "li",
  "blockquote",
  "a",
];
const ALLOWED_ATTR = ["href", "target", "rel"];

/**
 * Sanitizes announcement HTML before it is stored or rendered.
 *
 * Called from both the Zod schema (so a hand-built request never reaches
 * the database with anything unexpected in it) and every render site (so a
 * row written before this allow-list existed, or by a caller that skipped
 * the schema, still comes out safe).
 */
export function sanitizeAnnouncementHtml(html: string): string {
  return DOMPurify.sanitize(html, {
    ALLOWED_TAGS,
    ALLOWED_ATTR,
  }).trim();
}

/**
 * Strips markup down to plain text, collapsing the whitespace tags leave
 * behind. Used wherever the announcement's body needs to act like a plain
 * string again — search indexing, one-line table previews, and the
 * "did they actually type anything" check the submit buttons use, since an
 * empty rich-text document still serializes to `<p></p>`, never `""`.
 */
export function stripHtmlToText(html: string): string {
  return html
    .replace(/<[^>]*>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}
